Assets/Mirror/Runtime/Transport/Transport.cs

```diff
@@ -73,7 +74,23 @@ public virtual bool Available()
         /// as new channels</param>
         /// <param name="data">The data to send to the server</param>
         /// <returns>true if the send was successful</returns>
-        public abstract bool ClientSend(int channelId, byte[] data);
+        public virtual bool ClientSend(int channelId, byte[] data)
+        {
+            throw new NotImplementedException("Transports must override one of the ClientSend methods");
+        }
+
+        /// <summary>
+        /// Send a message to the server
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="channelId"></param>
+        /// <param name="msg"></param>
+        /// <returns></returns>
+        public virtual bool ClientSend<T>(int channelId, T msg) where T: IMessageBase
```

Comment:
why send<t> here?
why not just pass the byte[]?

Reply:
Because now the clients can send messages to the server allocation free by implementing this method

Reply:
why would it not work with just a byte[]?

Reply:
because there is no way to do allocation free messages if you use byte[]

Reply:
why not?

Reply:
because it forces mirror to allocate a byte[] of the right size

Reply:
didn't we have to allocate a byte[] of correct size before as well? as in: once per SendToAll etc

Reply:
yes vis,  the purpose of this is so that transport writers have a way to avoid it by overriding this method.

To be clear this PR per se does not fix any allocation,  but now there is a new ServerSend and ClientSend that does not use byte[]  that transports can override to provide allocation free sends.

After this PR there will be other PR's that remove allocations in some of the transports.

In Apathy you will override the new ServerSend and ClientSend.  By overriding them you can pass he data to apathy in an allocation free way (arraysegment or whatever else you want)